ddl/placement/constraint.go

```diff
@@ -0,0 +1,155 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package placement
+
+import (
+	"fmt"
+	"strings"
+)
+
+// ConstraintOp defines how a Constraint matches a store.
+type ConstraintOp string
+
+const (
+	// In restricts the store label value should in the value list.
+	// If label does not exist, `in` is always false.
+	In ConstraintOp = "in"
+	// NotIn restricts the store label value should not in the value list.
+	// If label does not exist, `notIn` is always true.
+	NotIn ConstraintOp = "notIn"
+	// Exists restricts the store should have the label.
+	Exists ConstraintOp = "exists"
+	// NotExists restricts the store should not have the label.
+	NotExists ConstraintOp = "notExists"
+)
+
+// Constraint is used to filter store when trying to place peer of a region.
+type Constraint struct {
+	Key    string       `json:"key,omitempty"`
+	Op     ConstraintOp `json:"op,omitempty"`
+	Values []string     `json:"values,omitempty"`
+}
+
+// NewConstraint will create a Constraint from string
+func NewConstraint(label string) (Constraint, error) {
+	r := Constraint{}
+
+	if len(label) < 4 {
+		return r, fmt.Errorf("%w: %s", ErrInvalidConstraintFormat, label)
+	}
+
+	var op ConstraintOp
+	switch label[0] {
+	case '+':
+		op = In
+	case '-':
+		op = NotIn
+	default:
+		return r, fmt.Errorf("%w: %s", ErrInvalidConstraintFormat, label)
+	}
+
+	kv := strings.Split(label[1:], "=")
+	if len(kv) != 2 {
+		return r, fmt.Errorf("%w: %s", ErrInvalidConstraintFormat, label)
+	}
+
+	key := strings.TrimSpace(kv[0])
+	if key == "" {
+		return r, fmt.Errorf("%w: %s", ErrInvalidConstraintFormat, label)
+	}
+
+	val := strings.TrimSpace(kv[1])
+	if val == "" {
+		return r, fmt.Errorf("%w: %s", ErrInvalidConstraintFormat, label)
+	}
+
+	if op == In && key == EngineLabelKey && strings.ToLower(val) == EngineLabelTiFlash {
+		return r, fmt.Errorf("%w: %s", ErrUnsupportedConstraint, label)
+	}
+
+	r.Key = key
+	r.Op = op
+	r.Values = []string{val}
+	return r, nil
+}
+
+// Restore converts a Constraint to a string.
+func (c *Constraint) Restore() (string, error) {
+	var sb strings.Builder
+	if len(c.Values) != 1 {
+		return "", fmt.Errorf("%w: constraint should have exactly one label value, got %v", ErrInvalidConstraintFormat, c.Values)
+	}
+	switch c.Op {
+	case In:
+		sb.WriteString("+")
+	case NotIn:
+		sb.WriteString("-")
+	default:
+		return "", fmt.Errorf("%w: disallowed operation '%s'", ErrInvalidConstraintFormat, c.Op)
+	}
+	sb.WriteString(c.Key)
+	sb.WriteString("=")
+	sb.WriteString(c.Values[0])
+	return sb.String(), nil
+}
+
+// String implements fmt.Stringer interface.
+func (c *Constraint) String() string {
+	str, err := c.Restore()
+	if err != nil {
+		return err.Error()
```

Comment:
The `String()` returning error message seems to be strange...

Reply:
Does panic feel better?

Reply:
How about log it and return empty(`""`)?

Reply:
That is too silent... The only way to construct invalid constraint is by manual construction, bypassing the new method and all other checks. So far, only tests and `json.Unmarshal` does not use the new method. `Unmarshal` takes data from the data persisted on PD, which are constructed through SQL and the new method.

User just should not go here. And devs should not rely on manual manipulation if there is a set of working public APIs.

Reply:
If it is not used in the normal case, we should not provide it. :thinking:

Reply:
But to use `json.Marshal` convert the struct to string, there must be one stringer implementation.

Reply:
Ok... Anyway, Golang cannot prohibits the developers from initializing a struct with manual construction. Panic can be dangerous in a production environment. I still prefer logging it and return an empty string:)

Reply:
We can prevent that by wrapping another internal package... Manual construction will be blocked unless you are in the placement package. E.g.:

```
type MyStruct struct {
   internal.MyRealStruct
}
```

But that is a little bit overkill to me.

In fact, one of my point for refactor is to prevent developers from manual/raw operations. I hope they can follow the contract by splitting packages.

Reply:
Anyway, I removed the stringer interface.